### DISCUSSION TITLE
Function Neumann BC Notation for Species Flux

### DISCUSSION THREAD

**Discussion by ghost** [shannon-lab/zapdos#104] [Q&A Plasma Transport] [answered]
How does one properly handle setting a BC for a species as a Neumann in the molar-log form? For example, let's say I want the following Neumann BC at the top of a 2D simulation, so that species are entering into the domain from the top:

[some_BC]
  type = ADFunctionNeumannBC
  ...
  function = 'BC_func'
[]
...
[BC_func]
  type = ParsedFunction
  value = '-1 * n * (a - b * x^2)'
[]

Where would I put the 'log( 1/ 6.022e23)', if at all, to put the value into molar-log form while conserving the flux & direction? Thanks!

**Comment by gsgall** [comment, marked as answer]
The way you can convert between the Zapdos value of a molar-log variable the density you want is with the expression 

n = exp( value ) * 6.022e23 

For example a conversion from a Zapdos value of -17.9135 translates to a density of 1e16. With that being said you will want to include the entire expression inside the logarithm so your expression would look like 

log(  -1 * n * ( a - b * x^2 ) / 6.022e23 ) 

For another example see the IC functions in this file 

https://github.com/shannon-lab/zapdos/blob/devel/test/tests/Lymberopoulos_rf_discharge/Lymberopoulos_with_argon_metastables.i

**Reply by ghost** [reply]
Wouldn't this be impossible to compute, since one would very likely end up trying to calculate the natural log of a negative number? Assuming a and b are positive such that (a - b * x^2) > 0.

**Reply by gsgall** [reply]
Part of the reason Zapdos uses the molar log form is to avoid negative densities. 
In this example appropriate a and b values should be selected such that 

-( a - b * x^2 ) > 0 

for all x in the domain of your mesh

**Reply by ghost** [reply]
Okay, then the expression inside the log must be positive, and the sign of the resulting term depends on if the expression at some given point x is below 1 or above 1. How does the system then differentiate between a negative and a positive 1st derivative (i.e. a species flux either entering or leaving a domain)?

**Reply by csdechant** [reply]
Hi, just to jump in, what @gsgall is saying about the density is Zapdos is true, that is the density is represented as n_0 = exp(N) * Na (where n_0 is the density in mol/m^3, N is the value calculated by Zapdos, and Na is Avogadro's number). So if you need to set a IC value or Dirichlet BC, one needs to use N = log(n_0 / Na). I believe the misunderstanding was that Neumann is a flux boundary condition. Fluxes in Zapdos are expressed normally (for example a flux ion BC of mu * -grad(V) * n_0 * normal is mu * -grad(V) * exp(N) * normal,  where mu is the ion mobility and V is the potential). So, in @speyres case, just make sure that the units of the NeumannBC is mol/(s*m^2) and you should be good.

@lindsayad, @cticenhour, @keniley1 if I over looked anything with this, please let me know.
